includes/class-wp-job-manager-post-types.php

```diff
@@ -390,14 +390,29 @@ public function register_post_types() {
 		 */
 		add_feed( self::get_job_feed_name(), [ $this, 'job_feed' ] );
 
+		/**
+		 * This code checks if (1) we're on a search or (2) the option to hide expired job listings is enabled.
+		 * If either condition is false, we're going to set the register_post_status() 'public' arg to false,
+		 * otherwise it will be true.
+		 *
+		 * This will prevent single expired job listings from showing a '404', instead of an
+		 * 'Expired' notice, when viewing the single job listing and not logged in.
+		 *
+		 * This will also address historical issues stemming from addressing the issue raised here
+		 * https://github.com/Automattic/wpjobmanager.com/issues/420.
+		 */
+		$is_site_search        = ( isset( $_GET['s'] ) );
+		$is_expired_searchable = (bool) get_option( 'job_manager_hide_expired' );
+		$is_expired_public     = ! $is_site_search || ! $is_expired_searchable;
+
 		/**
 		 * Post status
 		 */
 		register_post_status(
 			'expired',
 			[
 				'label'                     => _x( 'Expired', 'post status', 'wp-job-manager' ),
-				'public'                    => true,
+				'public'                    => $is_expired_public,
```

Comment:
@jom Here's the refactored implementation for setting `'public` on the post status registration. The logic is slightly different than discussed but noted in the comment. In testing:

- Job listings in WP Admin show and are counted as expected
- Expired Jobs are hidden from search results (or not, if the option is disabled)
- Expired Jobs, when viewed logged out, show the 'Expired' notice instead of a 404

Reply:
The variable name threw me off a bit, but I think the ultimate logic is correct!

Reply:
I ended up reverting this due to issues checking conditionals like `is_archive()`. Committed https://github.com/Automattic/WP-Job-Manager/pull/2449/commits/1a7b7da26d47c33255a23ea5e95da70773f6454a

Reply:
> This seems to be getting overwritten by the exclude_from_search parameter on the post-type registration:

I saw https://core.trac.wordpress.org/ticket/44737, but then saw the lines get changed in https://core.trac.wordpress.org/ticket/48556 and thought that would fix it, but now I can see what it is doing in the `} elseif ( ! $this->is_singular ) {` block. 

> I'm wondering if another option would be to keep the $query->set(), but instead of setting it to publish, retrieve all post statuses with get_post_stati(), remove what we don't want, and then pass everything else. Though, that seems a bit... hacky?

This is what I was going to suggest before I went down the 44737 rabbit hole because it seems like WP is doing this in a way that would work for us:
https://github.com/WordPress/WordPress/blob/32e94b4de108fb090ebc678b48113b8954f7f188/wp-includes/class-wp-query.php#L2587-L2593

I think if we did this we could just removed `expired` (I'm guessing)? It is a bit hack-y, but might be okay. 

The only other thought I had is modifying the global that holds the post status objects here after our conditionals and then restore `public => true` in the `posts_selection` action, but that might feel more hacky.

Reply:
@jom Here's the latest approach, which we talked about

https://github.com/Automattic/WP-Job-Manager/pull/2449/commits/82e75e67cb8136e5489f78f86872141e5155abbd